### ISSUE TITLE
Python console - Unadapted error ding when many auto-completion possibilities exist

### ISSUE THREAD

**Issue by CyrilleB79** [nvaccess/nvda#12631]
I open this issue because I have thought during a long time that auto-completion was not working well in the console.
The "ding" in case of multiple possibilities let me think that I was doing something wrong.

### Steps to reproduce:

* Press NVDA+control+Z to open the Python console
* type:
  `nav.__c`
* Press tab

### Actual behavior:

A ding is heard letting us think that pressing Tab in this case is not a valid action. Indeed this "ding" is the same as the Windows error ding, e.g. when you want to write something in a place where it is not allowed.
This makes the UX strange and makes the auto-completion feature not so easily self-discoverable.

### Expected behavior:

I can see two possibilities (to be discussed):
1. At the first tab press:
   * autocomplete if there is only one possibility (as today)
   * open the context menu if there is more than one possibility.
   I guess it is the standard way to do in IDEs. I haven't seen anywhere else this double tab to get the list of suggestions.
2. Use a beep rather than a ding. Indeed a beep has no special meaning whereas the ding indicates an error.

But other proposals are welcome.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
2021.1rc2

#### Windows version:
Windows 10 2004 (64-bit) build 19041.388
#### Name and version of other software in use when reproducing the issue:
N/A
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes, 2020.4: it was already the case.
#### If add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not tested

Cc @JulienCochuyt

**Comment by lukaszgo1** [comment]
>          I guess it is the standard way to do in IDEs. I haven't seen anywhere else this double tab to get the list of suggestions.

I always thought this has been done to make it similar to how tab auto-complete works in Bash i.e. tab auto-completes if there is only one possibility and if there are more the first tab does nothing and the second shows the list of suggestions.

**Comment by geoffshang** [comment]
> I always thought this has been done to make it similar to how tab auto-complete works in Bash i.e. tab auto-completes if there is only one possibility and if there are more the first tab does nothing and the second shows the list of suggestions.

At least in Bash, the first tab will complete as much as is unique.  So if 
there is only one match, it will complete it.  If there is more than one 
match, it will insert as much as it can, then stop.

So, for example, if you have two options ABCDE and ABC123 and you type AB, 
the first tab press will insert C and beep.

I'm not sure if the console works the same way but I'd be surprised if it 
doesn't.

Cheers,
Geoff.

**Comment by CyrilleB79** [comment]
I did not know about the bash example. But this makes sense and give an explanation. I am not using Unix/Linux regularly for a long time. And when I was using it many years ago, I do not remember such a behaviour with the beeps in the console (I was using csh or tcsh).

Anyway, I confirm this double tab behaviour in the Git bash consol under Windows. In the Git bash console however, there is no error ding. And of course no context menu but a list of suggestions and the prompt again.
In a real bash console (not the Windows Git bash), is there a "ding" or any other audio feedback at first tab in case of ambiguous auto-complete suggestion? Is this audio feedback something common? Does this audio feedback also sounds like an error sound?

Anyway I feel this way to do strange in the graphical Windows world, NVDA's python console being graphical.

I suggest one of these two changes:
1. Keep the "bash" way to do but:
   * change the ding to another beep, e.g. NVDA's `tones.beep` in order to avoid confusion with the Windows error "ding". This should allow people think that there is a dedicated feature here.
   * check the documentation and add more details if needed on this feature provided Windows developers may not be familiar with the Unix/Linux auto-complete feature.
2. Make something that looks more common to Windows developers. E.g. use only one tab pressure for auto-completion:
   * If there is only one possible auto-completion, auto-complete with the unique possible string
   * If there are two or more possibilities for auto-completion, auto-complete with the part common to all the possibilities and open a context menu with the possible suggestions

Note: Actually the Git bash and NVDA's Python console uses a 3 tab scheme:
* First tab (not always required): auto-complete as much as possible
* Second tab: ding
* Third tab: suggestions

**Comment by JulienCochuyt** [comment]
@CyrilleB79,
IMHO the current implementation is quite satisfying.
This triple action of the tab key for completion is indeed the default behavior of many Bash distributions, among other consoles.
There, it can often be customized to user preferences to different levels, but I guess such richness would be overkill in the context of the simple console provided by NVDA.

As for the choice of the "ding" sound: This is originally not meant to mean "error", but any kind of warning / notification / "something needs your attention". In Python, it is emitted as of calling `wx.Bell()`.
Btw, the audible bell of the MinTTY terminal shipped with Git Bash is disabled by default. Once reactivated, the same "ding" is emitted.

The second UX you describe is indeed more common to IDE users.
My suggestion would be not to change the tab-completion, but maybe add a `control+space` gesture to trigger this alternative behavior. Then, without the need for extra configuration, both console and IDE users would possibly feel more at ease.

**Comment by CyrilleB79** [comment]
@JulienCochuyt thanks for your comments. I will answer point by point to it:
1. I agree with you that customization of such a minor feature should not be implemented; it may create more confusion than resolve problems.
2. Regarding the ding, I maintain that it may be very misleading for Windows users, even for Git bash console users who did not dig into the option to activate the sound in the Git bash console. When I press a key in Windows and when I hear this sound, that is usually a signal that warns me that pressing again this key is useless because the expected action has not occurred, (e.g. cursor movement). I guess a `tones.beep` would be less confusing.
3. For my personal use, a control+space would be nice. But providing two ways to do the same thing would maybe be confusing for the users and for a such insignifiant feature. Let's wait for other people comments on this point if any.

To summarize I suggest the following implementation:
* Change `wx.Bell` to `tones.beep`.
* Describe with more details the auto-completion mechanism in the Dev Guide or even in the User Guide Console paragraph.
* Optionally if a vast majority of people agree, implement point 3., i.e. control+space alternative IDE-like shortcut.

What do you think?

**Comment by Adriani90** [comment]
@CyrilleB79 this issue is still relevant in NVDA 2025.1 Beta, correct?

**Comment by CyrilleB79** [comment]
Yes it is (NVDA 2025.1beta4)
